src/jit/compiler.h

```diff
@@ -3028,22 +3028,71 @@ class Compiler
     bool isIntrinsicAnIsSupportedPropertyGetter(NamedIntrinsic intrinsic);
     bool isFullyImplmentedISAClass(InstructionSet isa);
 #ifdef _TARGET_XARCH_
-    GenTree* impX86HWIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impSSEIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impSSE2Intrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impSSE3Intrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impSSSE3Intrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impSSE41Intrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impSSE42Intrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impAVXIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impAVX2Intrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impAESIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impBMI1Intrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impBMI2Intrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impFMAIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impLZCNTIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impPCLMULQDQIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
-    GenTree* impPOPCNTIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig);
+    GenTree* impUnsupportedHWIntrinsic(CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig, bool mustExpand);
+    GenTree* impX86HWIntrinsic(NamedIntrinsic        intrinsic,
+                               CORINFO_METHOD_HANDLE method,
+                               CORINFO_SIG_INFO*     sig,
+                               bool                  mustExpand);
+    GenTree* impSSEIntrinsic(NamedIntrinsic        intrinsic,
+                             CORINFO_METHOD_HANDLE method,
+                             CORINFO_SIG_INFO*     sig,
+                             bool                  mustExpand);
```

Comment:
Why do you pass `mustExpand` into `impXXXIntrinsic`, it seems only required by `impX86HWIntrinsic`.

Reply:
It is also used (in this PR) by Crc32, Lzcnt, and Popcnt, since they throw PNSE when varTypeIsLong(callType)

It will also need to be used by instructions like `Sse.Shuffle` to determine whether we should fallback to a GT_CALL or allow the instruction to be inlined (when op3 is not a constant).

Reply:
I think all the exception code can be unified to `impX86HWIntrinsic` with the table-driven approach. Now, perhaps you can directly extract the scalar intrinsic's exception code to `impX86HWIntrinsic`. Then I will improve it in the table-driven PR.

Reply:
We should probably do that particular refactoring with the table driven PR. The `varTypeIsLong` check shouldn't be for all instructions even after refactoring, so there isn't a good place to put it in `impX86HWIntrinsic` beforehand.

Reply:
OK.